## Background
One of the client, Category Manager for Chips, wants to have a better understanding on who their target customers are and what their purchasing behaviours are. As a data analyst in the Quantium's retail team, the goal for us is to get a supermarket's strategic plan for the chip category in the next half year by considering what metrics would be helpful to describe the customers' purchasing behavior. 

### Key
- customer segments 
- chip puchasing behavior



In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import plotly.express as px

In [2]:
trans_data = pd.read_excel('https://insidesherpa.s3.amazonaws.com/vinternships/companyassets/32A6DqtsbF7LbKdcq/QVI_transaction_data.xlsx')

In [3]:
trans_data.head(5)

DATE  STORE_NBR  LYLTY_CARD_NBR  TXN_ID  PROD_NBR  \
0  43390          1            1000       1         5   
1  43599          1            1307     348        66   
2  43605          1            1343     383        61   
3  43329          2            2373     974        69   
4  43330          2            2426    1038       108   

                                  PROD_NAME  PROD_QTY  TOT_SALES  
0    Natural Chip        Compny SeaSalt175g         2        6.0  
1                  CCs Nacho Cheese    175g         3        6.3  
2    Smiths Crinkle Cut  Chips Chicken 170g         2        2.9  
3    Smiths Chip Thinly  S/Cream&Onion 175g         5       15.0  
4  Kettle Tortilla ChpsHny&Jlpno Chili 150g         3       13.8

In [4]:
# Noticed that the format of the Date Column in transaction_data is off
# Changing to the right format
trans_data['DATE'] = pd.to_datetime(trans_data['DATE'], errors = 'coerce', unit = 'd', origin = '1900-01-01')
trans_data.head(5)

DATE  STORE_NBR  LYLTY_CARD_NBR  TXN_ID  PROD_NBR  \
0 2018-10-19          1            1000       1         5   
1 2019-05-16          1            1307     348        66   
2 2019-05-22          1            1343     383        61   
3 2018-08-19          2            2373     974        69   
4 2018-08-20          2            2426    1038       108   

                                  PROD_NAME  PROD_QTY  TOT_SALES  
0    Natural Chip        Compny SeaSalt175g         2        6.0  
1                  CCs Nacho Cheese    175g         3        6.3  
2    Smiths Crinkle Cut  Chips Chicken 170g         2        2.9  
3    Smiths Chip Thinly  S/Cream&Onion 175g         5       15.0  
4  Kettle Tortilla ChpsHny&Jlpno Chili 150g         3       13.8

In [5]:
# check to see if there are any missing values for the transaction data
trans_row_missing = trans_data.isnull().sum(axis=1)/trans_data.shape[1]
print(trans_row_missing[trans_row_missing>0.8])
# no missing values 

Series([], dtype: float64)


In [6]:
cust_data = pd.read_csv('https://insidesherpa.s3.amazonaws.com/vinternships/companyassets/32A6DqtsbF7LbKdcq/QVI_purchase_behaviour.csv')

In [7]:
cust_data.head(5)

LYLTY_CARD_NBR               LIFESTAGE PREMIUM_CUSTOMER
0            1000   YOUNG SINGLES/COUPLES          Premium
1            1002   YOUNG SINGLES/COUPLES       Mainstream
2            1003          YOUNG FAMILIES           Budget
3            1004   OLDER SINGLES/COUPLES       Mainstream
4            1005  MIDAGE SINGLES/COUPLES       Mainstream

In [8]:
# check to see if there are any missing values for the transaction data
cust_row_missing = cust_data.isnull().sum(axis=1)/cust_data.shape[1]
print(cust_row_missing[cust_row_missing>0.8])
# no missing values 

Series([], dtype: float64)


In [9]:
# Combine the 2 tables 
com_data = pd.merge(trans_data, cust_data, on = 'LYLTY_CARD_NBR', how = 'left')
com_data.head(5)

DATE  STORE_NBR  LYLTY_CARD_NBR  TXN_ID  PROD_NBR  \
0 2018-10-19          1            1000       1         5   
1 2019-05-16          1            1307     348        66   
2 2019-05-22          1            1343     383        61   
3 2018-08-19          2            2373     974        69   
4 2018-08-20          2            2426    1038       108   

                                  PROD_NAME  PROD_QTY  TOT_SALES  \
0    Natural Chip        Compny SeaSalt175g         2        6.0   
1                  CCs Nacho Cheese    175g         3        6.3   
2    Smiths Crinkle Cut  Chips Chicken 170g         2        2.9   
3    Smiths Chip Thinly  S/Cream&Onion 175g         5       15.0   
4  Kettle Tortilla ChpsHny&Jlpno Chili 150g         3       13.8   

                LIFESTAGE PREMIUM_CUSTOMER  
0   YOUNG SINGLES/COUPLES          Premium  
1  MIDAGE SINGLES/COUPLES           Budget  
2  MIDAGE SINGLES/COUPLES           Budget  
3  MIDAGE SINGLES/COUPLES           Budget  
4  MIDAGE SINGLES/COUPLES           Budget

In [10]:
# Only want the chip products in the table -- goal of our project 
chips_data = com_data[com_data['PROD_NAME'].str.contains('Chip') == True].copy()
chips_data.head(5)

DATE  STORE_NBR  LYLTY_CARD_NBR  TXN_ID  PROD_NBR  \
0 2018-10-19          1            1000       1         5   
2 2019-05-22          1            1343     383        61   
3 2018-08-19          2            2373     974        69   
6 2019-05-18          4            4149    3333        16   
8 2018-08-22          5            5026    4525        42   

                                  PROD_NAME  PROD_QTY  TOT_SALES  \
0    Natural Chip        Compny SeaSalt175g         2        6.0   
2    Smiths Crinkle Cut  Chips Chicken 170g         2        2.9   
3    Smiths Chip Thinly  S/Cream&Onion 175g         5       15.0   
6  Smiths Crinkle Chips Salt & Vinegar 330g         1        5.7   
8   Doritos Corn Chip Mexican Jalapeno 150g         1        3.9   

                LIFESTAGE PREMIUM_CUSTOMER  
0   YOUNG SINGLES/COUPLES          Premium  
2  MIDAGE SINGLES/COUPLES           Budget  
3  MIDAGE SINGLES/COUPLES           Budget  
6  MIDAGE SINGLES/COUPLES           Budget  
8  MIDAGE SINGLES/COUPLES           Budget

In [11]:
chips_data['PROD_NAME'].count()

74570

In [12]:
# Remove unnecessary information from the product name -- weight in this case
prod_name_split = chips_data["PROD_NAME"].str.replace(r'([0-9]+[gG])','').str.replace(r'[^\w]', ' ').str.split()
prod_name_split

0                   [Natural, Chip, Compny, SeaSalt]
2             [Smiths, Crinkle, Cut, Chips, Chicken]
3            [Smiths, Chip, Thinly, S, Cream, Onion]
6            [Smiths, Crinkle, Chips, Salt, Vinegar]
8           [Doritos, Corn, Chip, Mexican, Jalapeno]
                             ...                    
264815    [Cobs, Popd, Swt, Chlli, Sr, Cream, Chips]
264816                [Cobs, Popd, Sea, Salt, Chips]
264826                [Cobs, Popd, Sea, Salt, Chips]
264830                  [Thins, Chips, Light, Tangy]
264834      [Doritos, Corn, Chip, Mexican, Jalapeno]
Name: PROD_NAME, Length: 74570, dtype: object

### General Summary of the Products: 
To see what kinds of chips are most selling:
Thins chips
favour: Corn, Salt 

In [13]:
word_counts = {}

def count_words(line):
    for word in line:
        if word not in word_counts:
            word_counts[word] = 1
        else:
            word_counts[word] += 1
            
prod_name_split.apply(lambda line: count_words(line))
print(pd.Series(word_counts).sort_values(ascending=False))

Chips      49770
Corn       18878
Chip       18645
Smiths     16872
Doritos    15874
           ...  
Originl     1441
saltd       1441
CutSalt     1440
French      1418
Fries       1418
Length: 68, dtype: int64


In [14]:
chips_data.describe()

STORE_NBR  LYLTY_CARD_NBR         TXN_ID      PROD_NBR  \
count  74570.000000    7.457000e+04   74570.000000  74570.000000   
mean     135.457651    1.359095e+05  135533.580448     50.651039   
std       76.753058    8.025342e+04   77960.794363     32.792579   
min        1.000000    1.000000e+03       1.000000      1.000000   
25%       70.000000    7.013000e+04   68237.250000     27.000000   
50%      131.000000    1.311975e+05  135503.500000     42.000000   
75%      203.000000    2.032402e+05  203284.250000     78.000000   
max      272.000000    2.373711e+06  270209.000000    111.000000   

           PROD_QTY     TOT_SALES  
count  74570.000000  74570.000000  
mean       1.904264      6.860511  
std        0.343675      2.256661  
min        1.000000      1.900000  
25%        2.000000      5.800000  
50%        2.000000      6.600000  
75%        2.000000      7.800000  
max        5.000000     29.500000

In [15]:
chips_data.sort_values(by = 'PROD_QTY',ascending = False).head(5)

DATE  STORE_NBR  LYLTY_CARD_NBR  TXN_ID  PROD_NBR  \
69502  2018-08-20         55           55144   49328        44   
135230 2019-05-17         49           49361   46001        42   
81224  2019-05-19        217          217183  217138        44   
32670  2019-05-20        196          196102  196049         1   
135159 2018-08-20         19           19188   16426        93   

                                      PROD_NAME  PROD_QTY  TOT_SALES  \
69502            Thins Chips Light&  Tangy 175g         5       16.5   
135230  Doritos Corn Chip Mexican Jalapeno 150g         5       19.5   
81224            Thins Chips Light&  Tangy 175g         5       16.5   
32670   Smiths Crinkle Cut  Chips Barbecue 170g         5       14.5   
135159  Doritos Corn Chip Southern Chicken 150g         5       19.5   

                    LIFESTAGE PREMIUM_CUSTOMER  
69502          OLDER FAMILIES          Premium  
135230               RETIREES           Budget  
81224   OLDER SINGLES/COUPLES           Budget  
32670          OLDER FAMILIES           Budget  
135159               RETIREES           Budget

In [16]:
chips_data['DATE'].describe()

<ipython-input-16-3ad0a472c4c9>:1: FutureWarning: Treating datetime data as categorical rather than numeric in `.describe` is deprecated and will be removed in a future version of pandas. Specify `datetime_is_numeric=True` to silence this warning and adopt the future behavior now.
  chips_data['DATE'].describe()


count                   74570
unique                    364
top       2018-12-26 00:00:00
freq                      285
first     2018-07-03 00:00:00
last      2019-07-02 00:00:00
Name: DATE, dtype: object

### Summary of the transaction 
This table records the transaction history from July 3 2018 to July 2 2019, exactly 1 year. There are total 49770 transactions on chips purchases.

transaction_data['Brand Name'] = transaction_data['PROD_NAME'].str.split().apply(lambda x:x[0])
transaction_data['Brand Name'] = transaction_data['Brand Name'].replace('Red','RRD')
transaction_data['Brand Name'] = transaction_data['Brand Name'].replace('SNBTS','SUNBITES')
transaction_data['Brand Name'] = transaction_data['Brand Name'].replace('infzns','Infuzions')
transaction_data['Brand Name'] = transaction_data['Brand Name'].replace('WW','woolworths')
transaction_data['Brand Name'] = transaction_data['Brand Name'].replace('Smith','Smiths')
transaction_data['Brand Name'] = transaction_data['Brand Name'].replace('NCC','Natural')
transaction_data['Brand Name'] = transaction_data['Brand Name'].replace('Dorito','Doritos')
transaction_data['Brand Name'] = transaction_data['Brand Name'].replace('Grain','GrnWves')
transaction_data['Brand Name'].value_counts()

### Customer Behavior


In [18]:
# Overview on the customers information vs sales 
chips_group= pd.DataFrame(chips_data.groupby(["LIFESTAGE", "PREMIUM_CUSTOMER"])["TOT_SALES"].agg(["sum", "mean"]))
chips_group.round(0)

sum  mean
LIFESTAGE              PREMIUM_CUSTOMER               
MIDAGE SINGLES/COUPLES Budget             9838.0   7.0
                       Mainstream        23950.0   7.0
                       Premium           15349.0   7.0
NEW FAMILIES           Budget             5651.0   7.0
                       Mainstream         4307.0   7.0
                       Premium            3087.0   7.0
OLDER FAMILIES         Budget            44859.0   7.0
                       Mainstream        28298.0   7.0
                       Premium           21256.0   7.0
OLDER SINGLES/COUPLES  Budget            35943.0   7.0
                       Mainstream        35443.0   7.0
                       Premium           34545.0   7.0
RETIREES               Budget            30052.0   7.0
                       Mainstream        40592.0   7.0
                       Premium           24804.0   7.0
YOUNG FAMILIES         Budget            37064.0   7.0
                       Mainstream        25319.0   7.0
                       Premium           22781.0   7.0
YOUNG SINGLES/COUPLES  Budget            16777.0   6.0
                       Mainstream        40070.0   7.0
                       Premium           11599.0   6.0

In [19]:
# Get a bar graph to have a better understanding of the information above 
import dash
import dash_core_components as dcc
import dash_html_components as html
from dash.dependencies import Input, Output

In [20]:
chips_group_2 = chips_group.groupby(['LIFESTAGE','PREMIUM_CUSTOMER'])[['sum']].sum().reset_index()
chips_group_2

LIFESTAGE PREMIUM_CUSTOMER      sum
0   MIDAGE SINGLES/COUPLES           Budget   9838.3
1   MIDAGE SINGLES/COUPLES       Mainstream  23950.4
2   MIDAGE SINGLES/COUPLES          Premium  15349.4
3             NEW FAMILIES           Budget   5651.4
4             NEW FAMILIES       Mainstream   4307.1
5             NEW FAMILIES          Premium   3087.3
6           OLDER FAMILIES           Budget  44859.2
7           OLDER FAMILIES       Mainstream  28298.5
8           OLDER FAMILIES          Premium  21256.1
9    OLDER SINGLES/COUPLES           Budget  35943.0
10   OLDER SINGLES/COUPLES       Mainstream  35443.2
11   OLDER SINGLES/COUPLES          Premium  34545.0
12                RETIREES           Budget  30051.8
13                RETIREES       Mainstream  40592.1
14                RETIREES          Premium  24804.4
15          YOUNG FAMILIES           Budget  37064.1
16          YOUNG FAMILIES       Mainstream  25319.5
17          YOUNG FAMILIES          Premium  22781.1
18   YOUNG SINGLES/COUPLES           Budget  16777.1
19   YOUNG SINGLES/COUPLES       Mainstream  40069.9
20   YOUNG SINGLES/COUPLES          Premium  11599.4

In [23]:
fig = px.bar(chips_group_2, chips_group_2['LIFESTAGE'], chips_group_2['sum'], chips_group_2['PREMIUM_CUSTOMER'], text=(chips_group_2['sum']) )
fig.update_traces(texttemplate='%{text:.2s}', textposition='inside')
fig.update_layout(title='Customers Lifestage VS Total Sales (Jul 2018 - Jul 2019)',title_x=0.5,uniformtext_minsize=8, uniformtext_mode='hide')
fig.show()

In [26]:
# Overview on the customers information -- numbers of customers by membership status 
chips_member= pd.DataFrame(chips_data.groupby(["LIFESTAGE", "PREMIUM_CUSTOMER"])['LYLTY_CARD_NBR'].agg(["count", "mean"]))
chips_member.round(0)

count      mean
LIFESTAGE              PREMIUM_CUSTOMER                 
MIDAGE SINGLES/COUPLES Budget             1473  138353.0
                       Mainstream         3333  135685.0
                       Premium            2298  137792.0
NEW FAMILIES           Budget              840  134564.0
                       Mainstream          635  133854.0
                       Premium             449  133411.0
OLDER FAMILIES         Budget             6539  134077.0
                       Mainstream         4112  134891.0
                       Premium            3131  134714.0
OLDER SINGLES/COUPLES  Budget             5172  135251.0
                       Mainstream         5162  134362.0
                       Premium            4970  135888.0
RETIREES               Budget             4305  136330.0
                       Mainstream         5958  139730.0
                       Premium            3572  136775.0
YOUNG FAMILIES         Budget             5406  135824.0
                       Mainstream         3696  137776.0
                       Premium            3318  134706.0
YOUNG SINGLES/COUPLES  Budget             2708  136072.0
                       Mainstream         5643  134398.0
                       Premium            1850  139033.0

In [31]:
chips_member_2 = chips_member.groupby(['LIFESTAGE','PREMIUM_CUSTOMER'])[['count']].sum().reset_index()
chips_member_2

LIFESTAGE PREMIUM_CUSTOMER  count
0   MIDAGE SINGLES/COUPLES           Budget   1473
1   MIDAGE SINGLES/COUPLES       Mainstream   3333
2   MIDAGE SINGLES/COUPLES          Premium   2298
3             NEW FAMILIES           Budget    840
4             NEW FAMILIES       Mainstream    635
5             NEW FAMILIES          Premium    449
6           OLDER FAMILIES           Budget   6539
7           OLDER FAMILIES       Mainstream   4112
8           OLDER FAMILIES          Premium   3131
9    OLDER SINGLES/COUPLES           Budget   5172
10   OLDER SINGLES/COUPLES       Mainstream   5162
11   OLDER SINGLES/COUPLES          Premium   4970
12                RETIREES           Budget   4305
13                RETIREES       Mainstream   5958
14                RETIREES          Premium   3572
15          YOUNG FAMILIES           Budget   5406
16          YOUNG FAMILIES       Mainstream   3696
17          YOUNG FAMILIES          Premium   3318
18   YOUNG SINGLES/COUPLES           Budget   2708
19   YOUNG SINGLES/COUPLES       Mainstream   5643
20   YOUNG SINGLES/COUPLES          Premium   1850

In [40]:
fig2 = px.pie(chips_member_2, chips_member_2['PREMIUM_CUSTOMER'], chips_member_2['count'], color_discrete_sequence=px.colors.sequential.RdBu)
#fig.update_traces(textposition='inside', textinfo='percent+label')
fig2.update_layout(title='Percentage of the Customers - Membership',title_x=0.5,uniformtext_minsize=8, uniformtext_mode='hide')
fig2.show()

In [39]:
fig2 = px.pie(chips_member_2, chips_member_2['LIFESTAGE'], chips_member_2['count'], color_discrete_sequence=px.colors.sequential.RdBu)
#fig.update_traces(textposition='inside', textinfo='percent+label')
fig2.update_layout(title='Percentage of the Customers - Lifestage',title_x=0.5,uniformtext_minsize=8, uniformtext_mode='hide')
fig2.show()

## What can we tell from the graphs above

- Older Single/Couples customers has the highest total sales in combine, Retired customers comes 2nd and Older Familes customers comes 3rd 

To be more specific
- Older Families who are also budgetted are the top customers with 45K sales in total, retireed customers who are mainstreamed comes 2nd with 41K and young singles/couples comes 3rd with 40K

In general, 
- Most of our customers who purchased chips are elderly/retired or young families (reasons may be: (purchasing for themselves) more time to spend at home -> watching TV, (purchasing for others) have kids/youth at home, have parties/movies nights etc)
- premium members have less interest in purchasing chases (reasons may be: health diets, vegen etc)
-

Those groups should be our targetted customers 
